[WIP] Add loaded meshes to sharing link

daniel-wer opened the pull request:
- Includes some cleanup (review first commit separately to have an isolated look)
  - I was annoyed that because flow returns mixed for Object.values, we often had to add $FlowIssue comments, or worse, adapted the code to work around that by using Object.keys, converting the key to int and then mapping to get the values. I introduced `Utils.values` which returns the correct type and should be the only spot where the $FlowIssue comment is needed.
  - A couple of times I specified the type if the values were mapped, afterwards, which is a separate issue, since mapping `Array<number>` will lead to flow assuming a type of `number | any` for the mapped value argument :shrug: I did not systematically look for these occurrences, but I assume there are quite a few in our code, leading to less type security.
- 

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- abc

### Issues:
- fixes #

------
(Please delete unneeded items, merge only when none are left open)
- [ ] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Needs datastore update after deployment
- [ ] Ready for review

daniel-wer commented on frontend/javascripts/oxalis/model/sagas/isosurface_saga.js:
```diff
@@ -369,6 +358,164 @@ function* maybeLoadIsosurface(
+  });
+}
+
+function* loadPrecomputedMeshForSegmentId(
```
This function was moved from `segments_view_helper.js` and converted to a saga. Because of the race- and cancel-ability of sagas, the numerous checks whether the mesh was deleted in the mean time could be removed.

daniel-wer commented:
@philippotto I have mostly tested the happy paths for this PR, but will do some more monkey testing next week. Still, this should be ready for a first review :)

philippotto commented on frontend/javascripts/oxalis/model/sagas/isosurface_saga.js:
```diff
@@ -369,6 +374,168 @@ function* maybeLoadIsosurface(
+  });
+
+  try {
+    yield* call(processTaskWithPool, tasks, PARALLEL_PRECOMPUTED_MESH_LOADING_COUNT);
```
Are you sure that this function is cancelled once it was kicked off? I would have assumed that once an element from `tasks` has been started, that it's not cancelable by the redux sagas (a symptom of this is that the tasks are `async`s and not generator functions). The proper way to fix this would probably be to convert `processTaskWithPool` to be saga-based. It's not used anywhere else, so adapting callers wouldn't be a problem. However, adapting the tests could get a bit annoying..

A simple alternative would obviously be to re-introduce some of the "is canceled"-checks :see_no_evil:

daniel-wer replied:
Good point, I'll have a look and test what happens, but you're very likely correct.

> However, adapting the tests could get a bit annoying..

I don't think there are any tests for the mesh functionality yet, at least I haven't come across them :sweat_smile:
My plan is to add a screenshot test for this functionality. However, I'll also try to add a couple "higher-level" saga tests and will see whether it is too much "pain".

philippotto replied:
> I don't think there are any tests for the mesh functionality yet, at least I haven't come across them sweat_smile

I meant the unit tests which exist for `processTaskWithPool` :)

philippotto reviewed (changes requested):
Great work & refactoring :+1: It's a joy to read through these changes :man_dancing: 

I only have some concerns with `processTaskWithPool`. Maybe they are unfounded (I didn't test my assumption), but my current understanding is that this needs some changes.. Happy to call about this, if you want :call_me_hand:

daniel-wer commented on frontend/javascripts/oxalis/controller/url_manager.js:
```diff
@@ -181,6 +195,20 @@ class UrlManager {
+      if (currentMeshFile != null) {
+        const { meshFileName } = currentMeshFile;
+        const meshes = Utils.values(isosurfaces)
+          .filter(({ isVisible }: IsosurfaceInformation) => isVisible)
+          .map(({ segmentId, seedPosition, isPrecomputed }: IsosurfaceInformation) => ({
```
I added these types because of this observation:

> A couple of times I specified the type if the values were mapped, afterwards, which is a separate issue, since mapping `Array<number>` will lead to flow assuming a type of `number | any` for the mapped value argument :shrug: I did not systematically look for these occurrences, but I assume there are quite a few in our code, leading to less type security.

However, I revisited the type annotations and found out that the `any`s were not caused by the map, but instead because `stateByLayer` was initialized with `{}` without a type annotation. This led to flow inferring less-restrictive types. Adding the correct type annotation `UrlStateByLayer` to the initialization fixes this. So, as you say, these types can be removed and mapping doesn't produce `any`s :+1:

philippotto replied:
Ah, makes sense. Thanks for the explanation!

daniel-wer commented on frontend/javascripts/oxalis/view/action-bar/save_button.js:
```diff
@@ -17,6 +17,7 @@ import {
   LoadingOutlined,
 } from "@ant-design/icons";
 import ErrorHandling from "libs/error_handling";
+import { values } from "libs/utils";
```
Seems like a good idea :)

daniel-wer commented on frontend/javascripts/oxalis/model_initialization.js:
```diff
@@ -650,5 +655,24 @@ function applyLayerState(stateByLayer: UrlStateByLayer) {
+    if (layerState.meshInfo) {
+      const { meshFileName, meshes } = layerState.meshInfo;
+
+      Store.dispatch(updateCurrentMeshFileAction(effectiveLayerName, meshFileName));
```
Good that you mention it!
Currently, there is no way to find out which mesh file was used to load a precomputed mesh, after the fact. There is a `  +fileName?: string,` property in the `IsosurfaceInformation` in the store, but it doesn't seem to be populated, ever.

A related problem is which mapping was activated when an ad-hoc mesh was generated. Currently, this information is also not saved in any way.

I see two main directions we could take
(1) Save the `meshFileName` and `mappingName` information on a per-mesh basis and include it in the url. This would further increase the url length (maximum length for Edge [seems to be](https://stackoverflow.com/a/417184) 2047 characters).
(2) Revisit whether meshes should be invalidated/removed/hidden in certain situations (e.g.: if the mesh file is changed or the mapping is changed). Currently, the UI in the Segments tab can be pretty misleading. For example it is not possible to compute a mesh for segment 4 in the over segmentation, change the mapping to agglomerate_view_80 and compute a mesh for segment 4, there. I feel like we should clarify what should happen in these situations and then derive the url scheme from that, maybe?

Happy to jump on a call

philippotto replied:
Good points!

I think, we should explore direction (2) and see how we can handle these situations elegantly. However, I'd say this is better suited for a follow-up issue. Do you agree?

Approach (1) would probably work (and would be relatively simple, when ignoring URL limits), but the actual complexity (highlighted by (2) wouldn't be solved. So, it doesn't feel like the right way.

daniel-wer replied:
> I think, we should explore direction (2) and see how we can handle these situations elegantly. However, I'd say this is better suited for a follow-up issue. Do you agree?

Yes, I agree. So would you say, we keep the format as proposed in this PR for now (although it does not correctly work in some situations)?

Since we don't have a method to evolve the json url format to support older formats, yet, I would like to try to keep changes to a minimum, but it also doesn't feel right to merge this with known bugs :thinking:

daniel-wer commented on frontend/javascripts/oxalis/api/api_latest.js:
```diff
@@ -1654,7 +1653,9 @@ class DataApi {
     }
 
-    await loadMeshFromFile(segmentId, seedPosition, meshFileName, segmentationLayer, dataset);
+    Store.dispatch(
```
It is used and awaited in the connectome viewer (via the cross origin api). However, as far as I remember this stems from a time, where deleting a mesh while it was loading led to errors. Therefore, it is no longer relevant to await the mesh loading and might even speed up the connectome viewer, because both meshes will be loaded in parallel instead of sequentially.
Since the mesh loading can take quite long, I'm also not sure whether the blocking nature of this function would add much value. I'll adapt the usages and doc-string.

philippotto replied:
> It is used and awaited in the connectome viewer (via the cross origin api). However, as far as I remember this stems from a time, where deleting a mesh while it was loading led to errors. Therefore, it is no longer relevant to await the mesh loading and might even speed up the connectome viewer, because both meshes will be loaded in parallel instead of sequentially.

Two birds with one stone, I guess :tada: 

> Since the mesh loading can take quite long, I'm also not sure whether the blocking nature of this function would add much value. I'll adapt the usages and doc-string.

Sounds very plausible :+1:

philippotto commented on frontend/javascripts/libs/task_pool.js:
```diff
@@ -1,36 +1,50 @@
+  for (let i = 0; i < poolSize; i++) {
+    yield* fork(startNextTask);
+  }
+  // The saga will wait for all forked tasks to terminate before returning.
```
Once again, I'm noticing that redux saga is really neat :) I didn't know (or forgot?) the detail about attached vs unattached forks and was a bit surprised that this works that way (and that `spawn` is the counterpart to `fork`). I think, it wouldn't hurt to mention this detail in the comment, too. Then, the next reader might save one doc lookup :)

```suggestion
  // The saga will wait for all forked tasks to terminate before returning, because
  // fork() creates attached forks (in contrast to spawn()).
```

philippotto commented on frontend/javascripts/oxalis/model/sagas/isosurface_saga.js:
```diff
@@ -369,6 +374,172 @@ function* maybeLoadIsosurface(
+          chunkPosition,
+        );
+
+        const geometry = parseStlBuffer(stlData);
+        getSceneController().addIsosurfaceFromGeometry(geometry, id);
```
It probably doesn't make a real difference, since the parsing is synchronous, but using `call` for the two lines, would be more consistent, I guess.

philippotto reviewed (commented):
Great adaptation of the pool "saga" :1st_place_medal: I couldn't get the sharing link to work, yet, though. I tried it with adhoc meshes and the URL doesn't seem to contain anything mesh related :thinking:

philippotto commented on frontend/javascripts/oxalis/model/sagas/isosurface_saga.js:
```diff
@@ -369,6 +416,192 @@ function* maybeLoadIsosurface(
+    return;
+  }
+  // First create an array containing information about all loaded isosurfaces as the map is manipulated within the loop.
+  const isosurfacesMapForLayer = isosurfacesMapByLayer[segmentationLayer.name] || new Map();
```
If the `new Map` branch is hit, it probably should be stored in `isosurfacesMapByLayer`. Maybe something like this: 

```suggestion
  isosurfacesMapByLayer[segmentationLayer.name] = isosurfacesMapByLayer[segmentationLayer.name] || new Map();
  const isosurfacesMapForLayer = isosurfacesMapByLayer[segmentationLayer.name];
```

philippotto commented on frontend/javascripts/oxalis/model_initialization.js:
```diff
@@ -650,5 +655,36 @@ function applyLayerState(stateByLayer: UrlStateByLayer) {
+            loadAdHocMeshAction(
+              segmentId,
+              seedPosition,
+              // `undefined` values will be interpreted to mean the currently active mapping which
+              // is not desired for meshes loaded via url.
```
This seems a bit dangerous. Maybe the action creator should accept an optional `mappingInfo` object? If `mappingInfo == null`, the currently active mapping (if it exists) is used. If `mappingInfo` is defined, its content is used (which also allows `{mappingName: undefined, mappingName: undefined}` to enforce that no mapping is used). What do you think?

daniel-wer replied:
That's a very good idea and the code looks much better now :)

philippotto reviewed (commented):
I wasn't sure, whether the new round is ready for review again, but especially the last commit seemed like a bigger chunk ready to review :) Great work refactoring this! I only left two suggestions. Let me know when/whether I should also test again :)

daniel-wer commented:
> I wasn't sure, whether the new round is ready for review again, but especially the last commit seemed like a bigger chunk ready to review :)

Thanks a lot!

I addressed your feedback, would be great if you could do another test round.
I'll also test some more and will add a screenshot test as well.

philippotto reviewed (approved):
Sorry, this PR slipped through on my side! I reviewed the latest commits and also did another testing round. Looks very solid :+1: Looking forward to see this live :tada:

daniel-wer commented:
Nightly test ran through well: https://app.circleci.com/pipelines/github/scalableminds/webknossos/5717/workflows/22bbe067-6ff4-4a38-a907-c37a4132ee19/jobs/17004

philippotto commented on frontend/javascripts/test/puppeteer/dataset_rendering_helpers.js:
```diff
@@ -124,17 +127,29 @@ async function waitForTracingViewLoad(page: Page) {
-        threshold: 0.0,
-      });
+      changedPixels = pixelmatch(
+        // The buffers need to be converted to png before comparing them
```
Has this become necessary due to the update to `pixelmatch`? Or has this always been an issue? Good catch, either way :+1:

daniel-wer replied:
From what I read and experienced, `pixelmatch` started to assert this in the new version but the problem existed before. Usually the difference was fairly small (which is why it mostly worked before), but I also encountered garbled diffs during my testing with other `--use-gl` flags.

philippotto reviewed (approved):
Wonderful! Thank you so much for taking care of the screenshot tests :pray: I reviewed the latest commits and everything looks good!
